### PR TITLE
JAMES-3756 JMAP API endpoint should support delegation

### PR DESCRIPTION
**Pull request by quantranhong1999**
- [x] Delegation for JMAP API (QUAN)
 - [x] Delegate/set only be called by primaryAccount (QUAN)
 
Others (could be in diff PRs): 
 - [x] Delegation for upload (TUNG)
 - [x] Delegation for downloads (TUNG)
 - [x] Delegation for all kinds of push

### CODE REVIEW

**Review by chibenwa — APPROVED** [review]

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/SessionTranslator.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+   !hasAccountId(targetAccountId)(session.getUser)
+
+  private def hasAccountId(targetAccountId: AccountId)(username: Username): Boolean =
+    AccountId.from(username).toOption.contains(targetAccountId)
+
+  private def delegate(session: MailboxSession, targetAccountId: AccountId): SMono[MailboxSession] =
+    SFlux(delegationStore.delegatedUsers(session.getUser))
+      .filter(hasAccountId(targetAccountId))
+      .flatMap(targetUser => SMono.fromCallable(() => sessionProvider.authenticate(session.getUser).as(targetUser))
+        .subscribeOn(ReactorUtils.BLOCKING_CALL_WRAPPER))
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/SessionTranslator.scala:48]: Good catch!

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/SessionTranslator.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+class SessionTranslator  @Inject()(delegationStore: DelegationStore, sessionProvider: SessionProvider) {
+  def delegateIfNeeded(session: MailboxSession, targetAccountId: AccountId): SMono[MailboxSession] =
+    if (needDelegation(session, targetAccountId)) {
+      delegate(session, targetAccountId)
+    } else {
+      SMono.just(session)
+    }
+
+  private def needDelegation(session: MailboxSession, targetAccountId: AccountId): Boolean =
+   !hasAccountId(targetAccountId)(session.getUser)
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/SessionTranslator.scala:39]: Good catch
In the POC commit, it misses `!`

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/DelegateSetContract.scala [hunk 3]
```diff
@@ -516,4 +516,47 @@ trait DelegateSetContract {
...
+         |			}, "0"
+         |		]
+         |	]
+         |}""".stripMargin
+
+    val response =  `given`
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when
+      .post.prettyPeek()
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/DelegateSetContract.scala:544]: Debug

**Review by chibenwa — APPROVED** [review]

**Comment by quantranhong1999** [comment]
Rebased to solve conflict.

**Pull request merged**